core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -265,17 +277,32 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
             return allocationService.applyFailedShards(currentState, failedShards);
         }
 
-        private boolean shardExists(ClusterState currentState, ShardRoutingEntry task) {
+        private enum TaskClassification {
+            EXISTS_LEGAL,
+            EXISTS_ILLEGAL,
+            DOES_NOT_EXIST
+        }
+
+        private TaskClassification classifyTask(ClusterState currentState, ShardRoutingEntry task) {
```

Comment:
I find the names here a bit confusing - sorry :)  here is a suggestion I feel better about:

call this method `validateTask` (I think of it as a pre-flight validation) . the enum can be `ValidationResult`. The values are `SHARD_MISSING`, `SOURCE_INVALID`, `VALID`

wdyt?